File: config/security_monitoring.py
"""
Security Monitoring Configuration

Configuration settings for the security monitoring and logging system.
This file contains settings for anomaly detection rules, alert thresholds,
retention policies, and compliance requirements.

Task 23: Add security monitoring and logging
Requirements: SR1, SR2, SR5
"""

from typing import Dict, List, Any
from dataclasses import dataclass
from enum import Enum

class SecurityLevel(Enum):
    LOW = "Low"
    MEDIUM = "Medium"
    HIGH = "High"
    CRITICAL = "Critical"

class EventType(Enum):
    # Authentication events
    LOGIN_ATTEMPT = "LoginAttempt"
    LOGIN_SUCCESS = "LoginSuccess"
    LOGIN_FAILURE = "LoginFailure"
    TWO_FACTOR_REQUIRED = "TwoFactorRequired"
    TWO_FACTOR_SUCCESS = "TwoFactorSuccess"
    TWO_FACTOR_FAILURE = "TwoFactorFailure"
    SESSION_CREATED = "SessionCreated"
    SESSION_EXPIRED = "SessionExpired"
    SESSION_REVOKED = "SessionRevoked"
    ACCOUNT_LOCKED = "AccountLocked"
    ACCOUNT_UNLOCKED = "AccountUnlocked"
    
    # Transaction events
    BTC_COMMITMENT = "BTCCommitment"
    BTC_COMMITMENT_UPDATE = "BTCCommitmentUpdate"
    REWARD_CLAIM = "RewardClaim"
    PAYMENT_REQUEST = "PaymentRequest"
    PAYMENT_PROCESSED = "PaymentProcessed"
    PAYMENT_FAILED = "PaymentFailed"
    
    # Multisig events
    MULTISIG_PROPOSAL = "MultisigProposal"
    MULTISIG_SIGNING = "MultisigSigning"
    MULTISIG_EXECUTION = "MultisigExecution"
    KEY_ROTATION = "KeyRotation"
    EMERGENCY_MODE = "EmergencyMode"
    
    # KYC and compliance events
    KYC_SUBMISSION = "KYCSubmission"
    KYC_APPROVAL = "KYCApproval"
    KYC_REJECTION = "KYCRejection"
    COMPLIANCE_ALERT = "ComplianceAlert"
    ACCOUNT_FROZEN = "AccountFrozen"
    ACCOUNT_UNFROZEN = "AccountUnfrozen"
    AML_SCREENING = "AMLScreening"
    
    # Oracle and system events
    ORACLE_UPDATE = "OracleUpdate"
    ORACLE_FAILURE = "OracleFailure"
    SYSTEM_ERROR = "SystemError"
    SECURITY_VIOLATION = "SecurityViolation"
    
    # Anomaly detection events
    UNUSUAL_LOGIN_LOCATION = "UnusualLoginLocation"
    UNUSUAL_LOGIN_TIME = "UnusualLoginTime"
    HIGH_FREQUENCY_TRANSACTIONS = "HighFrequencyTransactions"
    LARGE_AMOUNT_TRANSACTION = "LargeAmountTransaction"
    SUSPICIOUS_PATTERN = "SuspiciousPattern"
    VELOCITY_ALERT = "VelocityAlert"
    DEVICE_CHANGE = "DeviceChange"
    IP_CHANGE = "IPChange"

@dataclass
class AnomalyRule:
    """Configuration for an anomaly detection rule"""
    rule_id: int
    name: str
    description: str
    event_types: List[EventType]
    enabled: bool
    threshold_value: float
    time_window_minutes: int
    severity: SecurityLevel
    auto_block: bool
    notification_required: bool

@dataclass
class SecurityMonitoringConfig:
    """Main configuration for security monitoring system"""
    
    # General settings
    enabled: bool = True
    retention_days: int = 365
    max_events_per_user: int = 1000
    auto_block_enabled: bool = True
    
    # Alert settings
    notification_webhook: str = None
    emergency_contacts: List[str] = None
    
    # Compliance settings
    compliance_retention_years: int = 10
    audit_trail_retention_years: int = 7
    
    # Performance settings
    max_concurrent_events: int = 100
    event_batch_size: int = 50
    cleanup_interval_hours: int = 24

# Default anomaly detection rules
DEFAULT_ANOMALY_RULES = [
    AnomalyRule(
        rule_id=1,
        name="Multiple Failed Logins",
        description="Detect multiple failed login attempts within a short time window",
        event_types=[EventType.LOGIN_FAILURE],
        enabled=True,
        threshold_value=5.0,
        time_window_minutes=15,
        severity=SecurityLevel.MEDIUM,
        auto_block=True,
        notification_required=True
    ),
    AnomalyRule(
        rule_id=2,
        name="High Value Transactions",
        description="Detect unusually large transactions that may indicate compromise",
        event_types=[EventType.BTC_COMMITMENT, EventType.REWARD_CLAIM],
        enabled=True,
        threshold_value=100000.0,  # $100k USD equivalent
        time_window_minutes=60,
        severity=SecurityLevel.HIGH,
        auto_block=False,
        notification_required=True
    ),
    AnomalyRule(
        rule_id=3,
        name="Rapid Transaction Velocity",
        description="Detect high frequency transactions that may indicate automated attacks",
        event_types=[EventType.PAYMENT_REQUEST, EventType.REWARD_CLAIM],
        enabled=True,
        threshold_value=10.0,
        time_window_minutes=5,
        severity=SecurityLevel.MEDIUM,
        auto_block=False,
        notification_required=True
    ),
    AnomalyRule(
        rule_id=4,
        name="Unusual Login Location",
        description="Detect logins from unusual geographic locations",
        event_types=[EventType.LOGIN_SUCCESS],
        enabled=True,
        threshold_value=1.0,
        time_window_minutes=1440,  # 24 hours
        severity=SecurityLevel.LOW,
        auto_block=False,
        notification_required=False
    ),
    AnomalyRule(
        rule_id=5,
        name="Security Violations",
        description="Detect any security violations that require immediate attention",
        event_types=[EventType.SECURITY_VIOLATION],
        enabled=True,
        threshold_value=1.0,
        time_window_minutes=1,
        severity=SecurityLevel.CRITICAL,
        auto_block=True,
        notification_required=True
    ),
    AnomalyRule(
        rule_id=6,
        name="2FA Bypass Attempts",
        description="Detect attempts to bypass two-factor authentication",
        event_types=[EventType.TWO_FACTOR_FAILURE],
        enabled=True,
        threshold_value=3.0,
        time_window_minutes=10,
        severity=SecurityLevel.HIGH,
        auto_block=True,
        notification_required=True
    ),
    AnomalyRule(
        rule_id=7,
        name="Emergency Mode Activation",
        description="Monitor emergency mode activations for potential security incidents",
        event_types=[EventType.EMERGENCY_MODE],
        enabled=True,
        threshold_value=1.0,
        time_window_minutes=1,
        severity=SecurityLevel.CRITICAL,
        auto_block=False,
        notification_required=True
    ),
    AnomalyRule(
        rule_id=8,
        name="Oracle Manipulation",
        description="Detect potential oracle manipulation attempts",
        event_types=[EventType.ORACLE_FAILURE],
        enabled=True,
        threshold_value=3.0,
        time_window_minutes=30,
        severity=SecurityLevel.HIGH,
        auto_block=False,
        notification_required=True
    ),
    AnomalyRule(
        rule_id=9,
        name="Compliance Alert Clustering",
        description="Detect clustering of compliance alerts that may indicate systematic issues",
        event_types=[EventType.COMPLIANCE_ALERT],
        enabled=True,
        threshold_value=5.0,
        time_window_minutes=60,
        severity=SecurityLevel.MEDIUM,
        auto_block=False,
        notification_required=True
    ),
    AnomalyRule(
        rule_id=10,
        name="Account Freeze Pattern",
        description="Monitor patterns of account freezes that may indicate targeted attacks",
        event_types=[EventType.ACCOUNT_FROZEN],
        enabled=True,
        threshold_value=3.0,
        time_window_minutes=120,
        severity=SecurityLevel.MEDIUM,
        auto_block=False,
        notification_required=True
    )
]

# Security level mappings for different event types
SECURITY_LEVEL_MAPPING = {
    # Critical events
    EventType.SECURITY_VIOLATION: SecurityLevel.CRITICAL,
    EventType.EMERGENCY_MODE: SecurityLevel.CRITICAL,
    
    # High severity events
    EventType.ACCOUNT_FROZEN: SecurityLevel.HIGH,
    EventType.KEY_ROTATION: SecurityLevel.HIGH,
    EventType.ORACLE_FAILURE: SecurityLevel.HIGH,
    EventType.LARGE_AMOUNT_TRANSACTION: SecurityLevel.HIGH,
    
    # Medium severity events
    EventType.LOGIN_FAILURE: SecurityLevel.MEDIUM,
    EventType.TWO_FACTOR_FAILURE: SecurityLevel.MEDIUM,
    EventType.COMPLIANCE_ALERT: SecurityLevel.MEDIUM,
    EventType.SUSPICIOUS_PATTERN: SecurityLevel.MEDIUM,
    EventType.HIGH_FREQUENCY_TRANSACTIONS: SecurityLevel.MEDIUM,
    EventType.UNUSUAL_LOGIN_LOCATION: SecurityLevel.MEDIUM,
    EventType.PAYMENT_FAILED: SecurityLevel.MEDIUM,
    
    # Low severity events (default for most events)
    EventType.LOGIN_SUCCESS: SecurityLevel.LOW,
    EventType.SESSION_CREATED: SecurityLevel.LOW,
    EventType.BTC_COMMITMENT: SecurityLevel.LOW,
    EventType.REWARD_CLAIM: SecurityLevel.LOW,
    EventType.PAYMENT_REQUEST: SecurityLevel.LOW,
    EventType.PAYMENT_PROCESSED: SecurityLevel.LOW,
    EventType.KYC_SUBMISSION: SecurityLevel.LOW,
    EventType.KYC_APPROVAL: SecurityLevel.LOW,
    EventType.ORACLE_UPDATE: SecurityLevel.LOW,
}

# Risk score calculation weights
RISK_SCORE_WEIGHTS = {
    "failed_login_attempts": 5,      # 5 points per failed login
    "suspicious_activity": 3,        # 3 points per suspicious activity
    "compliance_alerts": 5,          # 5 points per compliance alert
    "recent_suspicious_activity": 20, # 20 points if suspicious activity in last 7 days
    "kyc_tier_bonus": -10,          # -10 points for higher KYC tiers
    "account_age_bonus": -5,        # -5 points for accounts older than 6 months
}

# Risk score thresholds
RISK_SCORE_THRESHOLDS = {
    "low_risk": 0,
    "medium_risk": 30,
    "high_risk": 70,
    "critical_risk": 90
}

# Compliance settings
COMPLIANCE_SETTINGS = {
    "kyc_required_threshold": 100000,  # $100k USD equivalent
    "enhanced_dd_threshold": 500000,   # $500k USD equivalent
    "manual_review_threshold": 1000000, # $1M USD equivalent
    "auto_freeze_enabled": True,
    "screening_enabled": True,
    "retention_period_years": 10,
    "audit_trail_years": 7,
}

# Notification settings
NOTIFICATION_SETTINGS = {
    "webhook_enabled": False,
    "webhook_url": None,
    "email_enabled": False,
    "email_recipients": [],
    "sms_enabled": False,
    "sms_recipients": [],
    "slack_enabled": False,
    "slack_webhook": None,
}

# Performance and resource limits
PERFORMANCE_LIMITS = {
    "max_events_per_second": 100,
    "max_alerts_per_minute": 10,
    "max_concurrent_investigations": 5,
    "event_processing_timeout": 30,  # seconds
    "alert_processing_timeout": 60,  # seconds
    "cleanup_batch_size": 1000,
    "memory_limit_mb": 512,
}

# Audit trail settings
AUDIT_TRAIL_SETTINGS = {
    "enabled": True,
    "compliance_only": False,
    "include_state_changes": True,
    "include_metadata": True,
    "encrypt_sensitive_data": True,
    "retention_period_days": 365 * 7,  # 7 years
    "compliance_retention_days": 365 * 10,  # 10 years
    "cleanup_enabled": True,
    "cleanup_interval_hours": 24,
}

# Default configuration instance
DEFAULT_CONFIG = SecurityMonitoringConfig(
    enabled=True,
    retention_days=365,
    max_events_per_user=1000,
    auto_block_enabled=True,
    notification_webhook=None,
    emergency_contacts=[],
    compliance_retention_years=10,
    audit_trail_retention_years=7,
    max_concurrent_events=100,
    event_batch_size=50,
    cleanup_interval_hours=24
)

def get_security_level(event_type: EventType) -> SecurityLevel:
    """Get the security level for a given event type"""
    return SECURITY_LEVEL_MAPPING.get(event_type, SecurityLevel.LOW)

def calculate_risk_score(
    failed_logins: int,
    suspicious_activities: int,
    compliance_alerts: int,
    last_suspicious_timestamp: int = None,
    kyc_tier: int = 0,
    account_age_days: int = 0
) -> int:
    """Calculate risk score based on user behavior factors"""
    import time
    
    score = 0
    
    # Failed login attempts
    score += min(failed_logins * RISK_SCORE_WEIGHTS["failed_login_attempts"], 20)
    
    # Suspicious activity count
    score += min(suspicious_activities * RISK_SCORE_WEIGHTS["suspicious_activity"], 30)
    
    # Compliance alerts
    score += min(compliance_alerts * RISK_SCORE_WEIGHTS["compliance_alerts"], 30)
    
    # Recent suspicious activity
    if last_suspicious_timestamp:
        days_since = (int(time.time()) - last_suspicious_timestamp) // 86400
        if days_since < 7:
            score += RISK_SCORE_WEIGHTS["recent_suspicious_activity"]
        elif days_since < 30:
            score += RISK_SCORE_WEIGHTS["recent_suspicious_activity"] // 2
    
    # KYC tier bonus (higher tiers get lower risk scores)
    if kyc_tier > 0:
        score += RISK_SCORE_WEIGHTS["kyc_tier_bonus"] * kyc_tier
    
    # Account age bonus (older accounts get lower risk scores)
    if account_age_days > 180:  # 6 months
        score += RISK_SCORE_WEIGHTS["account_age_bonus"]
    
    return max(0, min(score, 100))

def get_risk_level(risk_score: int) -> str:
    """Get risk level based on risk score"""
    if risk_score >= RISK_SCORE_THRESHOLDS["critical_risk"]:
        return "critical"
    elif risk_score >= RISK_SCORE_THRESHOLDS["high_risk"]:
        return "high"
    elif risk_score >= RISK_SCORE_THRESHOLDS["medium_risk"]:
        return "medium"
    else:
        return "low"

def should_auto_block(event_type: EventType, risk_score: int) -> bool:
    """Determine if an event should trigger auto-blocking"""
    # Always block critical security violations
    if event_type == EventType.SECURITY_VIOLATION:
        return True
    
    # Block high-risk users for certain events
    if risk_score >= RISK_SCORE_THRESHOLDS["high_risk"]:
        high_risk_block_events = [
            EventType.LOGIN_FAILURE,
            EventType.TWO_FACTOR_FAILURE,
            EventType.LARGE_AMOUNT_TRANSACTION
        ]
        return event_type in high_risk_block_events
    
    return False

def get_retention_period(compliance_relevant: bool) -> int:
    """Get retention period in seconds based on compliance relevance"""
    if compliance_relevant:
        return COMPLIANCE_SETTINGS["retention_period_years"] * 365 * 24 * 3600
    else:
        return AUDIT_TRAIL_SETTINGS["retention_period_days"] * 24 * 3600

# Export configuration for use in other modules
__all__ = [
    'SecurityLevel',
    'EventType',
    'AnomalyRule',
    'SecurityMonitoringConfig',
    'DEFAULT_ANOMALY_RULES',
    'DEFAULT_CONFIG',
    'SECURITY_LEVEL_MAPPING',
    'RISK_SCORE_WEIGHTS',
    'RISK_SCORE_THRESHOLDS',
    'COMPLIANCE_SETTINGS',
    'NOTIFICATION_SETTINGS',
    'PERFORMANCE_LIMITS',
    'AUDIT_TRAIL_SETTINGS',
    'get_security_level',
    'calculate_risk_score',
    'get_risk_level',
    'should_auto_block',
    'get_retention_period'
]

File: programs/vault/src/state/security_monitoring.rs
use anchor_lang::prelude::*;
use std::collections::HashMap;

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug, PartialEq)]
pub enum SecurityEventType {
    // Authentication events
    LoginAttempt,
    LoginSuccess,
    LoginFailure,
    TwoFactorRequired,
    TwoFactorSuccess,
    TwoFactorFailure,
    SessionCreated,
    SessionExpired,
    SessionRevoked,
    AccountLocked,
    AccountUnlocked,
    
    // Transaction events
    BTCCommitment,
    BTCCommitmentUpdate,
    RewardClaim,
    PaymentRequest,
    PaymentProcessed,
    PaymentFailed,
    
    // Multisig events
    MultisigProposal,
    MultisigSigning,
    MultisigExecution,
    KeyRotation,
    EmergencyMode,
    
    // KYC and compliance events
    KYCSubmission,
    KYCApproval,
    KYCRejection,
    ComplianceAlert,
    AccountFrozen,
    AccountUnfrozen,
    AMLScreening,
    
    // Oracle and system events
    OracleUpdate,
    OracleFailure,
    SystemError,
    SecurityViolation,
    
    // Anomaly detection events
    UnusualLoginLocation,
    UnusualLoginTime,
    HighFrequencyTransactions,
    LargeAmountTransaction,
    SuspiciousPattern,
    VelocityAlert,
    DeviceChange,
    IPChange,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug, Copy)]
pub enum SecurityLevel {
    Low,
    Medium,
    High,
    Critical,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug, PartialEq)]
pub enum AlertStatus {
    Active,
    Investigating,
    Resolved,
    FalsePositive,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug)]
pub struct SecurityEvent {
    pub event_id: u64,
    pub event_type: SecurityEventType,
    pub user: Option<Pubkey>,
    pub timestamp: i64,
    pub ip_address: Option<String>,
    pub user_agent: Option<String>,
    pub device_id: Option<String>,
    pub session_id: Option<String>,
    pub transaction_id: Option<String>,
    pub amount: Option<u64>,
    pub details: String,
    pub metadata: HashMap<String, String>,
    pub security_level: SecurityLevel,
    pub requires_investigation: bool,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug)]
pub struct UserBehaviorProfile {
    pub user: Pubkey,
    pub created_at: i64,
    pub last_updated: i64,
    
    // Login patterns
    pub typical_login_hours: Vec<u8>, // Hours 0-23
    pub typical_login_days: Vec<u8>,  // Days 0-6 (Sunday-Saturday)
    pub common_locations: Vec<String>, // IP ranges or countries
    pub common_devices: Vec<String>,
    pub common_user_agents: Vec<String>,
    
    // Transaction patterns
    pub average_transaction_amount: u64,
    pub max_transaction_amount: u64,
    pub transaction_frequency: f64, // Transactions per day
    pub preferred_payment_methods: Vec<String>,
    
    // Risk indicators
    pub failed_login_attempts: u32,
    pub suspicious_activity_count: u32,
    pub last_suspicious_activity: Option<i64>,
    pub risk_score: u8, // 0-100
    pub is_high_risk: bool,
    
    // Compliance data
    pub kyc_tier: u8,
    pub compliance_alerts: u32,
    pub last_compliance_review: Option<i64>,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug)]
pub struct SecurityAlert {
    pub alert_id: u64,
    pub alert_type: SecurityEventType,
    pub user: Option<Pubkey>,
    pub created_at: i64,
    pub updated_at: i64,
    pub status: AlertStatus,
    pub security_level: SecurityLevel,
    pub description: String,
    pub related_events: Vec<u64>, // Event IDs
    pub investigation_notes: Vec<String>,
    pub assigned_to: Option<Pubkey>, // Security officer
    pub auto_resolved: bool,
    pub resolution_time: Option<i64>,
    pub false_positive: bool,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug)]
pub struct AnomalyDetectionRule {
    pub rule_id: u64,
    pub name: String,
    pub description: String,
    pub event_types: Vec<SecurityEventType>,
    pub enabled: bool,
    pub threshold_value: f64,
    pub time_window_minutes: u32,
    pub severity: SecurityLevel,
    pub auto_block: bool,
    pub notification_required: bool,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug)]
pub struct AuditTrail {
    pub trail_id: u64,
    pub user: Option<Pubkey>,
    pub action: String,
    pub resource: String,
    pub timestamp: i64,
    pub ip_address: Option<String>,
    pub user_agent: Option<String>,
    pub session_id: Option<String>,
    pub before_state: Option<String>, // JSON serialized state
    pub after_state: Option<String>,  // JSON serialized state
    pub success: bool,
    pub error_message: Option<String>,
    pub compliance_relevant: bool,
    pub retention_period: i64, // Seconds from creation
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug)]
pub struct SecurityMetrics {
    pub total_events: u64,
    pub events_by_type: HashMap<String, u64>,
    pub active_alerts: u64,
    pub resolved_alerts: u64,
    pub false_positives: u64,
    pub high_risk_users: u64,
    pub blocked_transactions: u64,
    pub average_resolution_time: f64, // Minutes
    pub last_updated: i64,
}

#[account]
pub struct SecurityMonitor {
    pub authority: Pubkey,
    pub event_counter: u64,
    pub alert_counter: u64,
    pub audit_counter: u64,
    pub enabled: bool,
    pub retention_days: u32,
    pub max_events_per_user: u32,
    pub auto_block_enabled: bool,
    pub notification_webhook: Option<String>,
    pub emergency_contacts: Vec<Pubkey>,
    pub created_at: i64,
    pub last_maintenance: i64,
}

#[account]
pub struct SecurityEventLog {
    pub monitor: Pubkey,
    pub events: Vec<SecurityEvent>,
    pub max_size: u32,
    pub created_at: i64,
    pub last_updated: i64,
}

#[account]
pub struct UserBehaviorStore {
    pub monitor: Pubkey,
    pub profiles: HashMap<Pubkey, UserBehaviorProfile>,
    pub created_at: i64,
    pub last_updated: i64,
}

#[account]
pub struct SecurityAlertStore {
    pub monitor: Pubkey,
    pub alerts: Vec<SecurityAlert>,
    pub active_count: u32,
    pub resolved_count: u32,
    pub created_at: i64,
    pub last_updated: i64,
}

#[account]
pub struct AnomalyRuleStore {
    pub monitor: Pubkey,
    pub rules: Vec<AnomalyDetectionRule>,
    pub enabled_count: u32,
    pub created_at: i64,
    pub last_updated: i64,
}

#[account]
pub struct AuditTrailStore {
    pub monitor: Pubkey,
    pub trails: Vec<AuditTrail>,
    pub compliance_trails: Vec<AuditTrail>,
    pub retention_policy: i64, // Seconds
    pub created_at: i64,
    pub last_cleanup: i64,
}

impl SecurityEvent {
    pub fn new(
        event_id: u64,
        event_type: SecurityEventType,
        user: Option<Pubkey>,
        details: String,
    ) -> Self {
        Self {
            event_id,
            event_type,
            user,
            timestamp: Clock::get().unwrap().unix_timestamp,
            ip_address: None,
            user_agent: None,
            device_id: None,
            session_id: None,
            transaction_id: None,
            amount: None,
            details,
            metadata: HashMap::new(),
            security_level: SecurityLevel::Low,
            requires_investigation: false,
        }
    }

    pub fn with_context(
        mut self,
        ip_address: Option<String>,
        user_agent: Option<String>,
        device_id: Option<String>,
        session_id: Option<String>,
    ) -> Self {
        self.ip_address = ip_address;
        self.user_agent = user_agent;
        self.device_id = device_id;
        self.session_id = session_id;
        self
    }

    pub fn with_transaction(mut self, transaction_id: String, amount: Option<u64>) -> Self {
        self.transaction_id = Some(transaction_id);
        self.amount = amount;
        self
    }

    pub fn with_security_level(mut self, level: SecurityLevel) -> Self {
        self.security_level = level;
        self.requires_investigation = matches!(level, SecurityLevel::High | SecurityLevel::Critical);
        self
    }

    pub fn add_metadata(mut self, key: String, value: String) -> Self {
        self.metadata.insert(key, value);
        self
    }
}

impl UserBehaviorProfile {
    pub fn new(user: Pubkey) -> Self {
        let now = Clock::get().unwrap().unix_timestamp;
        Self {
            user,
            created_at: now,
            last_updated: now,
            typical_login_hours: Vec::new(),
            typical_login_days: Vec::new(),
            common_locations: Vec::new(),
            common_devices: Vec::new(),
            common_user_agents: Vec::new(),
            average_transaction_amount: 0,
            max_transaction_amount: 0,
            transaction_frequency: 0.0,
            preferred_payment_methods: Vec::new(),
            failed_login_attempts: 0,
            suspicious_activity_count: 0,
            last_suspicious_activity: None,
            risk_score: 0,
            is_high_risk: false,
            kyc_tier: 0,
            compliance_alerts: 0,
            last_compliance_review: None,
        }
    }

    pub fn update_login_pattern(&mut self, hour: u8, day: u8, location: String, device: String, user_agent: String) {
        if !self.typical_login_hours.contains(&hour) {
            self.typical_login_hours.push(hour);
        }
        if !self.typical_login_days.contains(&day) {
            self.typical_login_days.push(day);
        }
        if !self.common_locations.contains(&location) {
            self.common_locations.push(location);
        }
        if !self.common_devices.contains(&device) {
            self.common_devices.push(device);
        }
        if !self.common_user_agents.contains(&user_agent) {
            self.common_user_agents.push(user_agent);
        }
        self.last_updated = Clock::get().unwrap().unix_timestamp;
    }

    pub fn update_transaction_pattern(&mut self, amount: u64, payment_method: String) {
        // Update average transaction amount
        if self.average_transaction_amount == 0 {
            self.average_transaction_amount = amount;
        } else {
            self.average_transaction_amount = (self.average_transaction_amount + amount) / 2;
        }

        // Update max transaction amount
        if amount > self.max_transaction_amount {
            self.max_transaction_amount = amount;
        }

        // Update preferred payment methods
        if !self.preferred_payment_methods.contains(&payment_method) {
            self.preferred_payment_methods.push(payment_method);
        }

        self.last_updated = Clock::get().unwrap().unix_timestamp;
    }

    pub fn calculate_risk_score(&mut self) -> u8 {
        let mut score = 0u8;

        // Failed login attempts (0-20 points)
        score += std::cmp::min(self.failed_login_attempts as u8 * 5, 20);

        // Suspicious activity count (0-30 points)
        score += std::cmp::min(self.suspicious_activity_count as u8 * 3, 30);

        // Recent suspicious activity (0-20 points)
        if let Some(last_suspicious) = self.last_suspicious_activity {
            let now = Clock::get().unwrap().unix_timestamp;
            let days_since = (now - last_suspicious) / 86400;
            if days_since < 7 {
                score += 20;
            } else if days_since < 30 {
                score += 10;
            }
        }

        // Compliance alerts (0-30 points)
        score += std::cmp::min(self.compliance_alerts as u8 * 5, 30);

        self.risk_score = std::cmp::min(score, 100);
        self.is_high_risk = self.risk_score >= 70;
        self.risk_score
    }

    pub fn is_anomalous_login(&self, hour: u8, day: u8, location: &str, device: &str) -> bool {
        let unusual_time = !self.typical_login_hours.contains(&hour) || !self.typical_login_days.contains(&day);
        let unusual_location = !self.common_locations.iter().any(|loc| loc.contains(location));
        let unusual_device = !self.common_devices.contains(&device.to_string());

        // Consider it anomalous if 2 or more factors are unusual
        [unusual_time, unusual_location, unusual_device].iter().filter(|&&x| x).count() >= 2
    }

    pub fn is_anomalous_transaction(&self, amount: u64) -> bool {
        // Transaction is anomalous if it's more than 5x the average or 2x the max
        amount > self.average_transaction_amount * 5 || amount > self.max_transaction_amount * 2
    }
}

impl SecurityAlert {
    pub fn new(
        alert_id: u64,
        alert_type: SecurityEventType,
        user: Option<Pubkey>,
        description: String,
        security_level: SecurityLevel,
    ) -> Self {
        let now = Clock::get().unwrap().unix_timestamp;
        Self {
            alert_id,
            alert_type,
            user,
            created_at: now,
            updated_at: now,
            status: AlertStatus::Active,
            security_level,
            description,
            related_events: Vec::new(),
            investigation_notes: Vec::new(),
            assigned_to: None,
            auto_resolved: false,
            resolution_time: None,
            false_positive: false,
        }
    }

    pub fn add_related_event(&mut self, event_id: u64) {
        if !self.related_events.contains(&event_id) {
            self.related_events.push(event_id);
        }
        self.updated_at = Clock::get().unwrap().unix_timestamp;
    }

    pub fn add_investigation_note(&mut self, note: String) {
        self.investigation_notes.push(note);
        self.updated_at = Clock::get().unwrap().unix_timestamp;
    }

    pub fn resolve(&mut self, false_positive: bool) {
        self.status = if false_positive { AlertStatus::FalsePositive } else { AlertStatus::Resolved };
        self.false_positive = false_positive;
        self.resolution_time = Some(Clock::get().unwrap().unix_timestamp);
        self.updated_at = Clock::get().unwrap().unix_timestamp;
    }

    pub fn assign_to(&mut self, officer: Pubkey) {
        self.assigned_to = Some(officer);
        self.status = AlertStatus::Investigating;
        self.updated_at = Clock::get().unwrap().unix_timestamp;
    }
}

impl AuditTrail {
    pub fn new(
        trail_id: u64,
        user: Option<Pubkey>,
        action: String,
        resource: String,
        success: bool,
    ) -> Self {
        Self {
            trail_id,
            user,
            action,
            resource,
            timestamp: Clock::get().unwrap().unix_timestamp,
            ip_address: None,
            user_agent: None,
            session_id: None,
            before_state: None,
            after_state: None,
            success,
            error_message: None,
            compliance_relevant: false,
            retention_period: 86400 * 365 * 7, // 7 years default
        }
    }

    pub fn with_context(
        mut self,
        ip_address: Option<String>,
        user_agent: Option<String>,
        session_id: Option<String>,
    ) -> Self {
        self.ip_address = ip_address;
        self.user_agent = user_agent;
        self.session_id = session_id;
        self
    }

    pub fn with_state_change(
        mut self,
        before_state: Option<String>,
        after_state: Option<String>,
    ) -> Self {
        self.before_state = before_state;
        self.after_state = after_state;
        self
    }

    pub fn with_error(mut self, error_message: String) -> Self {
        self.error_message = Some(error_message);
        self.success = false;
        self
    }

    pub fn mark_compliance_relevant(mut self) -> Self {
        self.compliance_relevant = true;
        self.retention_period = 86400 * 365 * 10; // 10 years for compliance
        self
    }
}
